### PR TITLE
Tests: Bid Adjustments Feature 

### PR DESCRIPTION
**Pull request by osulzhenko**
### 🔧 Type of changes
- [ ] new bid adapter
- [ ] update bid adapter
- [ ] new feature
- [ ] new analytics adapter
- [ ] new module
- [ ] bugfix
- [ ] documentation
- [ ] configuration
- [x] tech debt (test coverage, refactorings, etc.)

### ✨ What's the context?

What's the context for the changes? Are there any

### 🧠 Rationale behind the change

Why did you choose to make these changes? Were there any trade-offs you had to consider?

### 🔎 New Bid Adapter Checklist
- [ ] verify email contact works
- [ ] NO fully dynamic hosts
- [ ] geographic host parameters are NOT required
- [ ] NO direct use of HTTP is prohibited - *implement an existing Bidder interface that will do all the job*
- [ ] if the ORTB is just forwarded to the endpoint, use the generic adapter - *define the new adapter as the alias of the generic adapter*
- [ ] cover an adapter configuration with an integration test

### 🧪 Test plan

How do you know the changes are safe to ship to production?

### 🏎 Quality check

- [ ] Are your changes following [our code style guidelines](https://github.com/prebid/prebid-server-java/blob/master/docs/developers/code-style.md)?
- [ ] Are there any breaking changes in your code?
- [ ] Does your test coverage exceed 90%?
- [ ] Are there any erroneous console logs, debuggers or leftover code in your changes?

### CODE REVIEW

**Review by marki1an — CHANGES_REQUESTED** [review]
Add a few test cases for /amp endpoint  if it's possible

File: src/test/groovy/org/prebid/server/functional/model/request/auction/BidAdjustment.groovy [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+import com.fasterxml.jackson.databind.annotation.JsonNaming
+import groovy.transform.ToString
+import org.prebid.server.functional.util.PBSUtils
+
+@JsonNaming(PropertyNamingStrategies.LowerCaseStrategy)
+@ToString(includeNames = true, ignoreNulls = true)
+class BidAdjustment {
+
+    Map<BidAdjustmentMediaType, BidAdjustmentRule> mediaType
+    Integer version
```
[the remark this answers is not in the document]
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/model/request/auction/BidAdjustment.groovy:13]: https://github.com/prebid/prebid-server/issues/2085#:~:text=%22version%22%3A%201%2C%20%20//%20needed%20in%20case%20we%20extend%20to%20additional%20attributes%20someday

File: src/test/groovy/org/prebid/server/functional/tests/BidAdjustmentSpec.groovy [hunk 2]
```diff
@@ -125,14 +173,829 @@ class BidAdjustmentSpec extends BaseSpec {
...
+        assert response.seatbid.first.bid.first.price == getAdjustedPrice(originalPrice, ruleValue as BigDecimal, adjustmentType)
+        assert response.cur == bidResponse.cur
+
+        and: "Original bid price and currency should be presented in bid.ext"
+        verifyAll(response.seatbid.first.bid.first.ext) {
+            origbidcpm == originalPrice
+            origbidcur == bidResponse.cur
+        }
+
+        and: "Bidder request should contain default c"
```
Comment (marki1an) [inline src/test/groovy/org/prebid/server/functional/tests/BidAdjustmentSpec.groovy:216]: and: "Bidder request should contain default currency"

File: src/test/groovy/org/prebid/server/functional/tests/BidAdjustmentSpec.groovy [hunk 3]
```diff
@@ -125,14 +173,829 @@ class BidAdjustmentSpec extends BaseSpec {
...
+        verifyAll(response.seatbid.first.bid.first.ext) {
+            origbidcpm == originalPrice
+            origbidcur == bidResponse.cur
+        }
+
+        and: "Bidder request should contain currency from request"
+        def bidderRequest = bidder.getBidderRequest(bidRequest.id)
+        assert bidderRequest.cur == [currency]
+
+
```
Comment (marki1an) [inline src/test/groovy/org/prebid/server/functional/tests/BidAdjustmentSpec.groovy:943]: Please remove empty line

File: src/test/groovy/org/prebid/server/functional/tests/BidAdjustmentSpec.groovy [hunk 4]
```diff
@@ -125,14 +173,829 @@ class BidAdjustmentSpec extends BaseSpec {
...
+        when: "PBS processes auction request"
+        def response = pbsService.sendAuctionRequest(bidRequest)
+
+        then: "PBS should ignore bidAdjustments for this request"
+        assert response.seatbid.first.bid.first.price == originalPrice
+        assert response.cur == bidResponse.cur
+
+        and: "Should add a warning when in debug mode"
+        String errorMessage = "bid adjustment from request was invalid: the found rule [adjtype=${adjustmentType}, value=${adjustmentPrice}, currency=null] in banner.generic.* is invalid"
```
Comment (marki1an) [inline src/test/groovy/org/prebid/server/functional/tests/BidAdjustmentSpec.groovy:787]: deff errorMessage

File: src/test/groovy/org/prebid/server/functional/tests/BidAdjustmentSpec.groovy [hunk 5]
```diff
@@ -125,14 +173,829 @@ class BidAdjustmentSpec extends BaseSpec {
...
+        bidder.setResponse(bidRequest.id, bidResponse)
+
+        when: "PBS processes auction request"
+        def response = pbsService.sendAuctionRequest(bidRequest)
+
+        then: "PBS should ignore bidAdjustments for this request"
+        assert response.seatbid.first.bid.first.price == originalPrice
+        assert response.cur == bidResponse.cur
+
+        and: "Should add a warning when in debug mode"
```
Comment (marki1an) [inline src/test/groovy/org/prebid/server/functional/tests/BidAdjustmentSpec.groovy:786]: `PBs should emit logs`

File: src/test/groovy/org/prebid/server/functional/tests/BidAdjustmentSpec.groovy [hunk 6]
```diff
@@ -125,14 +173,829 @@ class BidAdjustmentSpec extends BaseSpec {
...
+        CPM            | MAX_CPM_ADJUST_VALUE        | NATIVE           | BidRequest.defaultNativeRequest
+        CPM            | MAX_CPM_ADJUST_VALUE        | ANY              | BidRequest.defaultBidRequest
+
+        STATIC         | MAX_STATIC_ADJUST_VALUE     | BANNER           | BidRequest.defaultBidRequest
+        STATIC         | MAX_STATIC_ADJUST_VALUE     | VIDEO_IN_STREAM  | BidRequest.defaultVideoRequest.tap { imp.first.video.placement = IN_STREAM }
+        STATIC         | MAX_STATIC_ADJUST_VALUE     | VIDEO_OUT_STREAM | BidRequest.defaultVideoRequest.tap { imp.first.video.placement = IN_ARTICLE }
+        STATIC         | MAX_STATIC_ADJUST_VALUE     | AUDIO            | BidRequest.defaultAudioRequest
+        STATIC         | MAX_STATIC_ADJUST_VALUE     | NATIVE           | BidRequest.defaultNativeRequest
+        STATIC         | MAX_CPM_ADJUST_VALUE        | ANY              | BidRequest.defaultBidRequest
+    }
```
Comment (marki1an) [inline src/test/groovy/org/prebid/server/functional/tests/BidAdjustmentSpec.groovy:242]: The last one should be
`STATIC         | MAX_STATIC_ADJUST_VALUE        | ANY              | BidRequest.defaultBidRequest`
Add just one test case with the average max rule value.

File: src/test/groovy/org/prebid/server/functional/tests/BidAdjustmentSpec.groovy [hunk 7]
```diff
@@ -125,14 +173,829 @@ class BidAdjustmentSpec extends BaseSpec {
...
+        }
+        bidder.setResponse(bidRequest.id, bidResponse)
+
+        when: "PBS processes auction request"
+        def response = pbsService.sendAuctionRequest(bidRequest)
+
+        then: "Final bid price should be adjusted"
+        def convertedAdjustment = convertCurrency(adjustmentRule.value, adjustmentRule.currency, bidResponse.cur)
+        def adjustedBidPrice = getAdjustedPrice(originalPrice, convertedAdjustment, adjustmentRule.adjustmentType)
+        response.seatbid.first.bid.first.price == convertCurrency(adjustedBidPrice, bidResponse.cur, bidRequest.cur.first)
```
Comment (marki1an) [inline src/test/groovy/org/prebid/server/functional/tests/BidAdjustmentSpec.groovy:541]: please add assert key word before

**Review by marki1an — CHANGES_REQUESTED** [review]

**Review by marki1an — APPROVED** [review]

**Pull request merged**